{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "scdefg.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Munfred/scdefg/blob/main/scdefg.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Hm9pzUCAR8Ej"
      },
      "source": [
        "# last altered jan 22 01:09"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "iiyR_cbZnXWW",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "769fd050-95c4-41d6-a3f6-42cc25487317"
      },
      "source": [
        "# clone repo which contains the html templates for flask\n",
        "!git clone https://github.com/Munfred/scdefg.git"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'scdefg'...\n",
            "remote: Enumerating objects: 40, done.\u001b[K\n",
            "remote: Counting objects: 100% (40/40), done.\u001b[K\n",
            "remote: Compressing objects: 100% (34/34), done.\u001b[K\n",
            "remote: Total 40 (delta 11), reused 10 (delta 4), pack-reused 0\u001b[K\n",
            "Unpacking objects: 100% (40/40), done.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1cljkrUaPeyg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "aaadfa65-1301-422b-e542-d3339ec913a1"
      },
      "source": [
        "# install flask dependencies on colab, takes a minute or two\n",
        "!pip install scvi-tools --quiet\n",
        "!pip install Flask-Misaka --quiet\n",
        "!pip install Flask-Markdown --quiet\n",
        "!pip install flask-ngrok --quiet\n",
        "!pip install flask==0.12.2 --quiet\n",
        "# Newer versions of flask don't work in Colab\n",
        "# See https://github.com/plotly/dash/issues/257"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\u001b[K     |████████████████████████████████| 153kB 6.1MB/s \n",
            "\u001b[K     |████████████████████████████████| 194kB 8.4MB/s \n",
            "\u001b[K     |████████████████████████████████| 245kB 8.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 122kB 9.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 51kB 4.7MB/s \n",
            "\u001b[K     |████████████████████████████████| 133kB 6.2MB/s \n",
            "\u001b[?25h  Building wheel for misaka (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[K     |████████████████████████████████| 174kB 7.0MB/s \n",
            "\u001b[?25h  Building wheel for Flask-Markdown (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[K     |████████████████████████████████| 92kB 4.6MB/s \n",
            "\u001b[?25h"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "v1ZYxoLvaxBb",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a9c07fce-01aa-4436-bd89-ccfed7c9df18"
      },
      "source": [
        "from flask import Flask, Response, jsonify, request, render_template, Blueprint, send_file, redirect, url_for\n",
        "import logging\n",
        "import pandas as pd\n",
        "import sys\n",
        "import json\n",
        "import time\n",
        "from io import StringIO\n",
        "import urllib\n",
        "from flaskext.markdown import Markdown\n",
        "from flask_misaka import Misaka\n",
        "from flask_ngrok import run_with_ngrok\n",
        "from google.colab import files\n",
        "import anndata\n",
        "import scvi\n",
        "import numpy as np\n",
        "from tqdm import tqdm\n",
        "print('scvi-tools verseion:', scvi.__version__)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "scvi-tools verseion: 0.8.1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xxiwiCB4GgQg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bc2540c3-e9c2-409e-fd1b-c237d67aa376"
      },
      "source": [
        "# this cell will download the file taylor2020.h5ad with the C. elegans single cell \n",
        "# data from the CeNGEN project. The data is described in the preprint \n",
        "# \"Molecular topography of an entire nervous system\", by Taylor et al 2020\n",
        "# https://doi.org/10.1016/j.neuron.2018.07.042\n",
        "\n",
        "!wget -nc https://github.com/Munfred/wormcells-site/releases/download/taylor2020/taylor2020.h5ad\n",
        "\n",
        "# this is an example, you can just as easily load your own .h5ad file"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-01-22 03:24:35--  https://github.com/Munfred/wormcells-site/releases/download/taylor2020/taylor2020.h5ad\n",
            "Resolving github.com (github.com)... 140.82.113.4\n",
            "Connecting to github.com (github.com)|140.82.113.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 Moved Permanently\n",
            "Location: https://github.com/Munfred/wormcells-data/releases/download/taylor2020/taylor2020.h5ad [following]\n",
            "--2021-01-22 03:24:35--  https://github.com/Munfred/wormcells-data/releases/download/taylor2020/taylor2020.h5ad\n",
            "Reusing existing connection to github.com:443.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/222185132/0b5c6880-5149-11eb-8695-20a4f0843b75?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210122%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210122T032435Z&X-Amz-Expires=300&X-Amz-Signature=c4eb1c9315ac989690d024ebca1abb6d1e0525aa6b0f9891799a407455257906&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=222185132&response-content-disposition=attachment%3B%20filename%3Dtaylor2020.h5ad&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-01-22 03:24:35--  https://github-production-release-asset-2e65be.s3.amazonaws.com/222185132/0b5c6880-5149-11eb-8695-20a4f0843b75?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210122%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210122T032435Z&X-Amz-Expires=300&X-Amz-Signature=c4eb1c9315ac989690d024ebca1abb6d1e0525aa6b0f9891799a407455257906&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=222185132&response-content-disposition=attachment%3B%20filename%3Dtaylor2020.h5ad&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.153.84\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.153.84|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 381516791 (364M) [application/octet-stream]\n",
            "Saving to: ‘taylor2020.h5ad’\n",
            "\n",
            "taylor2020.h5ad     100%[===================>] 363.84M  62.6MB/s    in 7.0s    \n",
            "\n",
            "2021-01-22 03:24:42 (52.1 MB/s) - ‘taylor2020.h5ad’ saved [381516791/381516791]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZED3NGP9GgNR",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 439
        },
        "outputId": "b2633a1f-6d46-43d2-c88b-7fe1ebe656a5"
      },
      "source": [
        "adata = anndata.read('taylor2020.h5ad')\n",
        "print(adata)\n",
        "display(adata.obs.head())\n",
        "display(adata.var.head())\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "AnnData object with n_obs × n_vars = 100955 × 46911\n",
            "    obs: 'dropbox_id', 'counts', 'experiment_code', 'cell_type', 'tissue'\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dropbox_id</th>\n",
              "      <th>counts</th>\n",
              "      <th>experiment_code</th>\n",
              "      <th>cell_type</th>\n",
              "      <th>tissue</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGAGACGAA</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>65</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>Unannotated</td>\n",
              "      <td>Unannotated</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGGTAAACT</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>367</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>AVF</td>\n",
              "      <td>Neuron</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGGTAGCCA</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>1792</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>AVH</td>\n",
              "      <td>Neuron</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGTAACCCT</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>1229</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>RIA</td>\n",
              "      <td>Neuron</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1806-ST-1-AAACCTGAGTACGCGA</th>\n",
              "      <td>1806-ST-1</td>\n",
              "      <td>1401</td>\n",
              "      <td>Pan-1</td>\n",
              "      <td>AUA</td>\n",
              "      <td>Neuron</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                           dropbox_id  counts  ...    cell_type       tissue\n",
              "1806-ST-1-AAACCTGAGAGACGAA  1806-ST-1      65  ...  Unannotated  Unannotated\n",
              "1806-ST-1-AAACCTGAGGTAAACT  1806-ST-1     367  ...          AVF       Neuron\n",
              "1806-ST-1-AAACCTGAGGTAGCCA  1806-ST-1    1792  ...          AVH       Neuron\n",
              "1806-ST-1-AAACCTGAGTAACCCT  1806-ST-1    1229  ...          RIA       Neuron\n",
              "1806-ST-1-AAACCTGAGTACGCGA  1806-ST-1    1401  ...          AUA       Neuron\n",
              "\n",
              "[5 rows x 5 columns]"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>gene_id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WBGene00014450</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00014451</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010957</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010958</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00014452</th>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: []\n",
              "Index: [WBGene00014450, WBGene00014451, WBGene00010957, WBGene00010958, WBGene00014452]"
            ]
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "mZ1mTnWlPlMw",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "befd9688-d071-4de5-9220-a8991139c965"
      },
      "source": [
        "adata.obs[adata.obs['experiment_code']==experiment]['cell_type'].value_counts()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Epidermis    360\n",
              "AIZ          326\n",
              "Germline     249\n",
              "AIM          231\n",
              "URA          195\n",
              "            ... \n",
              "DVA            0\n",
              "AVD            0\n",
              "AVB            0\n",
              "RIM            0\n",
              "M3             0\n",
              "Name: cell_type, Length: 169, dtype: int64"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 63
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-CrUMiF7LoHr",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 467
        },
        "outputId": "ec12068f-d6a1-4ebe-8af9-b35598358c1b"
      },
      "source": [
        "# our adata has annotations by experiment, by cell type and by tissue\n",
        "# we use the experiment code and cell type to generate a dataframe\n",
        "# that counts how many cells of each type are present in each experiment\n",
        "# this dataframe is needed to create the table that allows the user to select \n",
        "# which cells to compare on the app\n",
        "\n",
        "# cell type does on the index, experiment on the columns, entries are number of\n",
        "# cells of that type per experiment\n",
        "\n",
        "experiments= np.sort(adata.obs.experiment_code.unique())\n",
        "unique_cell_types=np.sort(adata.obs.cell_type.unique())\n",
        "census=pd.DataFrame(np.nan, index= unique_cell_types, columns=experiments)\n",
        "\n",
        "for experiment in experiments:\n",
        "    census[experiment] = census.index.map(adata.obs[adata.obs['experiment_code']==experiment]['cell_type'].value_counts())\n",
        "\n",
        "\n",
        "census.index=census.index.rename('Cell Type')\n",
        "census.to_csv('census.csv')\n",
        "census"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Pan-1</th>\n",
              "      <th>Pan-2</th>\n",
              "      <th>acr-2</th>\n",
              "      <th>ceh-28_dat-1</th>\n",
              "      <th>ceh-34</th>\n",
              "      <th>cho-1_1</th>\n",
              "      <th>cho-1_2</th>\n",
              "      <th>eat-4</th>\n",
              "      <th>ift-20</th>\n",
              "      <th>nlp-13_ceh-2</th>\n",
              "      <th>nmr-1</th>\n",
              "      <th>tph-1_ceh-10</th>\n",
              "      <th>unc-3</th>\n",
              "      <th>unc-47_1</th>\n",
              "      <th>unc-47_2</th>\n",
              "      <th>unc-53</th>\n",
              "      <th>unc-86</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Cell Type</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ADA</th>\n",
              "      <td>91</td>\n",
              "      <td>130</td>\n",
              "      <td>10</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>4</td>\n",
              "      <td>115</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>63</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>3</td>\n",
              "      <td>174</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ADE</th>\n",
              "      <td>48</td>\n",
              "      <td>70</td>\n",
              "      <td>0</td>\n",
              "      <td>282</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>50</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ADF</th>\n",
              "      <td>3</td>\n",
              "      <td>16</td>\n",
              "      <td>12</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>7</td>\n",
              "      <td>243</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>464</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ADL</th>\n",
              "      <td>8</td>\n",
              "      <td>14</td>\n",
              "      <td>11</td>\n",
              "      <td>5</td>\n",
              "      <td>7</td>\n",
              "      <td>6</td>\n",
              "      <td>7</td>\n",
              "      <td>330</td>\n",
              "      <td>512</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>28</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>12</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AFD</th>\n",
              "      <td>27</td>\n",
              "      <td>57</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>141</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>VD_DD</th>\n",
              "      <td>125</td>\n",
              "      <td>153</td>\n",
              "      <td>39</td>\n",
              "      <td>30</td>\n",
              "      <td>6</td>\n",
              "      <td>5</td>\n",
              "      <td>8</td>\n",
              "      <td>11</td>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>5</td>\n",
              "      <td>5</td>\n",
              "      <td>31</td>\n",
              "      <td>594</td>\n",
              "      <td>464</td>\n",
              "      <td>31</td>\n",
              "      <td>12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Vulval_cells</th>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>13</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>56</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Vulval_muscle</th>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>5</td>\n",
              "      <td>11</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>20</td>\n",
              "      <td>2</td>\n",
              "      <td>9</td>\n",
              "      <td>5</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>42</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>XXX</th>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>12</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>hmc</th>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "      <td>10</td>\n",
              "      <td>2</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>9</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>169 rows × 17 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "               Pan-1  Pan-2  acr-2  ...  unc-47_2  unc-53  unc-86\n",
              "Cell Type                           ...                          \n",
              "ADA               91    130     10  ...         5       3     174\n",
              "ADE               48     70      0  ...         0       2       0\n",
              "ADF                3     16     12  ...         0       1       1\n",
              "ADL                8     14     11  ...         5      12       4\n",
              "AFD               27     57      2  ...         3       2       0\n",
              "...              ...    ...    ...  ...       ...     ...     ...\n",
              "VD_DD            125    153     39  ...       464      31      12\n",
              "Vulval_cells       3      3      1  ...         0       0      10\n",
              "Vulval_muscle      2      3      5  ...         1      42       9\n",
              "XXX                1      1      2  ...         1       0       1\n",
              "hmc                2      1      4  ...         0       1       1\n",
              "\n",
              "[169 rows x 17 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 73
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "XVuldf52LX7o",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4fe3a1b2-be03-4812-d560-c7533a3d132d"
      },
      "source": [
        "adata.obs.experiment_code.unique()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['Pan-1', 'Pan-2', 'unc-47_1', 'nmr-1', 'unc-47_2', ..., 'ceh-34', 'unc-86', 'unc-53', 'nlp-13_ceh-2', 'ceh-28_dat-1']\n",
              "Length: 17\n",
              "Categories (17, object): ['Pan-1', 'Pan-2', 'unc-47_1', 'nmr-1', ..., 'unc-86', 'unc-53',\n",
              "                          'nlp-13_ceh-2', 'ceh-28_dat-1']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 39
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "I8tlE5P7Q6UP",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 473
        },
        "outputId": "a846d3c2-4e5c-4fa0-a4c9-c62e583e7fbb"
      },
      "source": [
        "pd.read_csv('./scdefg/df.csv')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Cell Type</th>\n",
              "      <th>L2_experiment_1</th>\n",
              "      <th>L2_experiment_2</th>\n",
              "      <th>Waterston_300_minutes</th>\n",
              "      <th>Waterston_400_minutes</th>\n",
              "      <th>Waterston_500_minutes_batch_1</th>\n",
              "      <th>Waterston_500_minutes_batch_2</th>\n",
              "      <th>Murray_r17</th>\n",
              "      <th>Murray_b01</th>\n",
              "      <th>Murray_b02</th>\n",
              "      <th>acr-2</th>\n",
              "      <th>ceh-34</th>\n",
              "      <th>cho-1_1</th>\n",
              "      <th>cho-1_2</th>\n",
              "      <th>ift-20</th>\n",
              "      <th>eat-4</th>\n",
              "      <th>unc-47_2</th>\n",
              "      <th>unc-47_1</th>\n",
              "      <th>nmr-1</th>\n",
              "      <th>Pan</th>\n",
              "      <th>tph-1_ceh-10</th>\n",
              "      <th>unc-3</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ABarpaaa_lineage</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>92</td>\n",
              "      <td>96</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>41</td>\n",
              "      <td>28</td>\n",
              "      <td>14</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ADA?</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>7</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>124</td>\n",
              "      <td>4</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>168</td>\n",
              "      <td>1</td>\n",
              "      <td>62</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ADA_ADE_grandparent</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>6</td>\n",
              "      <td>32</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>7</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ADE</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>46</td>\n",
              "      <td>16</td>\n",
              "      <td>9</td>\n",
              "      <td>9</td>\n",
              "      <td>6</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>50</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>109</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ADE_CEP_PDE</td>\n",
              "      <td>126</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>337</th>\n",
              "      <td>Vulval_precursor_bin_1_late_L2</td>\n",
              "      <td>282</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>338</th>\n",
              "      <td>Vulval_precursor_bin_2_early_L3</td>\n",
              "      <td>271</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>339</th>\n",
              "      <td>XXX</td>\n",
              "      <td>73</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>6</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>8</td>\n",
              "      <td>6</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>340</th>\n",
              "      <td>Y</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>18</td>\n",
              "      <td>6</td>\n",
              "      <td>10</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>341</th>\n",
              "      <td>Z1_Z4</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>43</td>\n",
              "      <td>110</td>\n",
              "      <td>39</td>\n",
              "      <td>40</td>\n",
              "      <td>31</td>\n",
              "      <td>82</td>\n",
              "      <td>27</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>342 rows × 22 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                           Cell Type  L2_experiment_1  ...  tph-1_ceh-10  unc-3\n",
              "0                   ABarpaaa_lineage                0  ...             0      0\n",
              "1                               ADA?                0  ...             1     62\n",
              "2                ADA_ADE_grandparent                0  ...             0      0\n",
              "3                                ADE                0  ...             1      2\n",
              "4                        ADE_CEP_PDE              126  ...             0      0\n",
              "..                               ...              ...  ...           ...    ...\n",
              "337   Vulval_precursor_bin_1_late_L2              282  ...             0      0\n",
              "338  Vulval_precursor_bin_2_early_L3              271  ...             0      0\n",
              "339                              XXX               73  ...             0      0\n",
              "340                                Y                0  ...             0      0\n",
              "341                            Z1_Z4                0  ...             0      0\n",
              "\n",
              "[342 rows x 22 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 75
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Cd-Fi9LBQhjM",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 436
        },
        "outputId": "a8fe977a-ed80-42c0-9af4-b26e6d27e4f1"
      },
      "source": [
        "df = pd.read_csv('./scdefg/df.csv')\n",
        "df"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Cell Type</th>\n",
              "      <th>Pan-1</th>\n",
              "      <th>Pan-2</th>\n",
              "      <th>acr-2</th>\n",
              "      <th>ceh-28_dat-1</th>\n",
              "      <th>ceh-34</th>\n",
              "      <th>cho-1_1</th>\n",
              "      <th>cho-1_2</th>\n",
              "      <th>eat-4</th>\n",
              "      <th>ift-20</th>\n",
              "      <th>nlp-13_ceh-2</th>\n",
              "      <th>nmr-1</th>\n",
              "      <th>tph-1_ceh-10</th>\n",
              "      <th>unc-3</th>\n",
              "      <th>unc-47_1</th>\n",
              "      <th>unc-47_2</th>\n",
              "      <th>unc-53</th>\n",
              "      <th>unc-86</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ADA</td>\n",
              "      <td>91</td>\n",
              "      <td>130</td>\n",
              "      <td>10</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>4</td>\n",
              "      <td>115</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>63</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>3</td>\n",
              "      <td>174</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ADE</td>\n",
              "      <td>48</td>\n",
              "      <td>70</td>\n",
              "      <td>0</td>\n",
              "      <td>282</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>50</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ADF</td>\n",
              "      <td>3</td>\n",
              "      <td>16</td>\n",
              "      <td>12</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>7</td>\n",
              "      <td>243</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>464</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ADL</td>\n",
              "      <td>8</td>\n",
              "      <td>14</td>\n",
              "      <td>11</td>\n",
              "      <td>5</td>\n",
              "      <td>7</td>\n",
              "      <td>6</td>\n",
              "      <td>7</td>\n",
              "      <td>330</td>\n",
              "      <td>512</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>28</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>12</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>AFD</td>\n",
              "      <td>27</td>\n",
              "      <td>57</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>141</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>164</th>\n",
              "      <td>VD_DD</td>\n",
              "      <td>125</td>\n",
              "      <td>153</td>\n",
              "      <td>39</td>\n",
              "      <td>30</td>\n",
              "      <td>6</td>\n",
              "      <td>5</td>\n",
              "      <td>8</td>\n",
              "      <td>11</td>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>5</td>\n",
              "      <td>5</td>\n",
              "      <td>31</td>\n",
              "      <td>594</td>\n",
              "      <td>464</td>\n",
              "      <td>31</td>\n",
              "      <td>12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>165</th>\n",
              "      <td>Vulval_cells</td>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>13</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>56</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>166</th>\n",
              "      <td>Vulval_muscle</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>5</td>\n",
              "      <td>11</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>20</td>\n",
              "      <td>2</td>\n",
              "      <td>9</td>\n",
              "      <td>5</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>42</td>\n",
              "      <td>9</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>167</th>\n",
              "      <td>XXX</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>12</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>168</th>\n",
              "      <td>hmc</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "      <td>10</td>\n",
              "      <td>2</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>9</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>169 rows × 18 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "         Cell Type  Pan-1  Pan-2  acr-2  ...  unc-47_1  unc-47_2  unc-53  unc-86\n",
              "0              ADA     91    130     10  ...         0         5       3     174\n",
              "1              ADE     48     70      0  ...         0         0       2       0\n",
              "2              ADF      3     16     12  ...         0         0       1       1\n",
              "3              ADL      8     14     11  ...         0         5      12       4\n",
              "4              AFD     27     57      2  ...         0         3       2       0\n",
              "..             ...    ...    ...    ...  ...       ...       ...     ...     ...\n",
              "164          VD_DD    125    153     39  ...       594       464      31      12\n",
              "165   Vulval_cells      3      3      1  ...         0         0       0      10\n",
              "166  Vulval_muscle      2      3      5  ...         1         1      42       9\n",
              "167            XXX      1      1      2  ...         1         1       0       1\n",
              "168            hmc      2      1      4  ...         0         0       1       1\n",
              "\n",
              "[169 rows x 18 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 74
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4P4EEdYiGgDF",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 422
        },
        "outputId": "7c7f16d4-d982-4c82-ffbd-c1b3904fd182"
      },
      "source": [
        "# in order to create the interactive volcano plot with gene names and descriptions\n",
        "# we download the C. elegans gene descriptions from the Alliance of Genome Resources\n",
        "!wget -nc http://download.alliancegenome.org/3.2.0/GENE-DESCRIPTION-TSV/WB/GENE-DESCRIPTION-TSV_WB_10.tsv.gz\n",
        "# then we parse the gene descriptions into a dataframe and add those to the adata\n",
        "# the alliance download page at https://www.alliancegenome.org/downloads  provides\n",
        "# gene descriptions for C.elegans, zebrafish, drosophilam humans, rat, mouse and yeast\n",
        "# so by tweaking the code in this cell you can add gene descriptions to your own dataset!\n",
        "gene_descriptions=pd.read_csv('GENE-DESCRIPTION-TSV_WB_10.tsv.gz', sep='\\t', skiprows=14, header=None)\n",
        "gene_descriptions.columns=['orig_id','gene_name', 'gene_description']\n",
        "gene_descriptions['gene_id']=gene_descriptions['orig_id'].str.split(':',expand=True)[1]\n",
        "gene_descriptions.index=gene_descriptions['gene_id'].values\n",
        "\n",
        "# we add the gene descriptions to adata.var by mapping them using the gene id that\n",
        "# is already in the adata.var index\n",
        "adata.var['gene_id']=adata.var.index.map(gene_descriptions['gene_id'])\n",
        "adata.var['gene_name']=adata.var.index.map(gene_descriptions['gene_name'])\n",
        "adata.var['gene_description']=adata.var.index.map(gene_descriptions['gene_description'])\n",
        "adata.var.head()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-01-22 03:37:11--  http://download.alliancegenome.org/3.2.0/GENE-DESCRIPTION-TSV/WB/GENE-DESCRIPTION-TSV_WB_10.tsv.gz\n",
            "Resolving download.alliancegenome.org (download.alliancegenome.org)... 13.249.134.109, 13.249.134.49, 13.249.134.3, ...\n",
            "Connecting to download.alliancegenome.org (download.alliancegenome.org)|13.249.134.109|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 885437 (865K) [application/x-gzip]\n",
            "Saving to: ‘GENE-DESCRIPTION-TSV_WB_10.tsv.gz’\n",
            "\n",
            "GENE-DESCRIPTION-TS 100%[===================>] 864.68K  --.-KB/s    in 0.06s   \n",
            "\n",
            "2021-01-22 03:37:11 (15.3 MB/s) - ‘GENE-DESCRIPTION-TSV_WB_10.tsv.gz’ saved [885437/885437]\n",
            "\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene_id</th>\n",
              "      <th>gene_name</th>\n",
              "      <th>gene_description</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>gene_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WBGene00014450</th>\n",
              "      <td>WBGene00014450</td>\n",
              "      <td>MTCE.1</td>\n",
              "      <td>No description available</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00014451</th>\n",
              "      <td>WBGene00014451</td>\n",
              "      <td>MTCE.2</td>\n",
              "      <td>No description available</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010957</th>\n",
              "      <td>WBGene00010957</td>\n",
              "      <td>nduo-6</td>\n",
              "      <td>Predicted to have NADH dehydrogenase (ubiquino...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010958</th>\n",
              "      <td>WBGene00010958</td>\n",
              "      <td>ndfl-4</td>\n",
              "      <td>Predicted to have NADH dehydrogenase (ubiquino...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00014452</th>\n",
              "      <td>WBGene00014452</td>\n",
              "      <td>MTCE.5</td>\n",
              "      <td>No description available</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                       gene_id  ...                                   gene_description\n",
              "gene_id                         ...                                                   \n",
              "WBGene00014450  WBGene00014450  ...                           No description available\n",
              "WBGene00014451  WBGene00014451  ...                           No description available\n",
              "WBGene00010957  WBGene00010957  ...  Predicted to have NADH dehydrogenase (ubiquino...\n",
              "WBGene00010958  WBGene00010958  ...  Predicted to have NADH dehydrogenase (ubiquino...\n",
              "WBGene00014452  WBGene00014452  ...                           No description available\n",
              "\n",
              "[5 rows x 3 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 38
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WNy2ckXxJ_mS"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "fKJ6uVc6Gf4I"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aRG30muRgbdy",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0c12989f-c0d5-4e7a-8884-2096c62aed42"
      },
      "source": [
        "logging.basicConfig(level=logging.INFO)\n",
        "logger=logging.getLogger(__name__)\n",
        "logger.info('Starting wormcells-de...')\n",
        "\n",
        "flask_app = Flask(__name__, template_folder='./scdefg/templates', static_folder='./scdefg/static')\n",
        "flask_app.config['JSONIFY_PRETTYPRINT_REGULAR'] = False\n",
        "\n",
        "run_with_ngrok(flask_app)\n",
        "Misaka(flask_app, math_explicit = True)\n",
        "\n",
        "\n",
        "tables = Blueprint('tables', __name__, url_prefix='/tables')\n",
        "\n",
        "# read in the census dataframe with the number of cells of each label in each dataset\n",
        "# to render the table titles better we replace underscores with spaces,\n",
        "# and use non breaking hyphens (&#8209;) \n",
        "df_nice_names = pd.read_csv(flask_app.open_resource('census.csv'))\n",
        "df_nice_names.columns = df_nice_names.columns.str.replace('_',' ')\n",
        "df_nice_names.columns = df_nice_names.columns.str.replace('-','&#8209;')\n",
        "\n",
        "# convert df to dict for sending as json to datatables\n",
        "dict_df = df_nice_names.to_dict(orient='records')\n",
        "# convert column names into dict for sending as json to datatables\n",
        "columns = [{\"data\": item, \"title\": item} for item in df_nice_names.columns]\n",
        "\n",
        "#### datatables to render the selection tables ####\n",
        "@tables.route(\"/\", methods=['GET'])\n",
        "def clientside_table_content():\n",
        "    return jsonify({'data': dict_df, 'columns': columns})\n",
        "\n",
        "flask_app.register_blueprint(tables)\n",
        "\n",
        "# this is the landing page\n",
        "@flask_app.route(\"/\")\n",
        "def clientside_table():\n",
        "    return render_template(\"clientside_table.html\")\n",
        "\n",
        "\n",
        "####\n",
        "\n",
        "@flask_app.route(\"/test\")\n",
        "def test():\n",
        "    return render_template(\"test.html\")\n",
        "\n",
        "# @flask_app.route(\"/\")\n",
        "# def index():\n",
        "#     logger.info('Got a request for index!')\n",
        "#     return render_template(\"index.html\")\n",
        "\n",
        "@flask_app.route('/submit', methods=['POST', 'GET'])\n",
        "def receive_submission():\n",
        "    logger.info('Got a submission!')\n",
        "    print('PRESSED THE SUBMISSION BUTTOOOON')\n",
        "    # answer is a dict of json strings containing selected row and column index numbers\n",
        "    answer = request.form.to_dict(flat=False)\n",
        "    print(answer)\n",
        "    print(df.head())\n",
        "\n",
        "    \n",
        "    # need to convert the json strings to dict, then to a data frame\n",
        "    # data1 is the selection for the first group, data2 for the second\n",
        "    data1 = json.loads(answer['data1'][0])\n",
        "    data1_df = pd.DataFrame.from_dict(data1[0])\n",
        "\n",
        "    print(data1_df)\n",
        "    data2 = json.loads(answer['data2'][0])\n",
        "    data2_df = pd.DataFrame.from_dict(data2[0])\n",
        "\n",
        "    # now map the index number to experiment name and cell type name\n",
        "    group1_df = pd.DataFrame()\n",
        "    group1_df['cell_type1'] = data1_df['row'].map(df['Cell Type'])\n",
        "    group1_df['experiment1'] = data1_df['column'].map(pd.Series(df.columns.values))\n",
        "\n",
        "    group2_df = pd.DataFrame()\n",
        "    group2_df['cell_type2'] = data2_df['row'].map(df['Cell Type'])\n",
        "    group2_df['experiment2'] = data2_df['column'].map(pd.Series(df.columns.values))\n",
        "\n",
        "    genes = StringIO(json.loads(answer['genes'][0]))\n",
        "    genes_df = pd.read_csv(genes, names=['selected_genes'])\n",
        "\n",
        "    jobname_df = pd.DataFrame(columns=['job_name'])\n",
        "    jobname_df = jobname_df.append({'job_name': 'bla'}, ignore_index=True)\n",
        "\n",
        "    selected_groups_df = pd.concat([group1_df, group2_df, genes_df, jobname_df], axis=1)\n",
        "    print(selected_groups_df)\n",
        "\n",
        "    csv_buffer = StringIO()\n",
        "    selected_groups_df.to_csv(csv_buffer)\n",
        "    # files.download(\"readme.md\")\n",
        "\n",
        "\n",
        "    # return 'derpderp'\n",
        "    print('RETURN FILE NOW')\n",
        "    # return send_file('/content/df.csv')\n",
        "    # return send_file('./foo.txt', as_attachment=True)\n",
        "    return redirect(url_for('test'))\n",
        "\n",
        "@flask_app.route(\"/get_results_csv\")\n",
        "def get_results_csv():\n",
        "    # csv = pd.read_csv('results.csv')\n",
        "    with open(\"results.csv\") as fp:\n",
        "        csv = fp.read()  \n",
        "    # csv = '1,2,3\\n4,5,6\\n'\n",
        "    return Response(\n",
        "        csv,\n",
        "        mimetype=\"text/csv\",\n",
        "        headers={\"Content-disposition\":\n",
        "                 \"attachment; filename=results.csv\"})\n",
        "\n",
        "@flask_app.route('/file_downloads')\n",
        "def file_downloads():\n",
        "\ttry:\n",
        "\t\treturn render_template(\"\"\"\n",
        "\t\t\t<body class=\"body\">\n",
        "\t\t\t\t<div class=\"container\" align=\"left\">\n",
        "\t\t\t\t\t<a href=\"/return_files/\" target=\"blank\"><button class='btn btn-default'>Download!</button></a>\n",
        "\t\t\t\t</div>\n",
        "\t\t\t</body>\t\n",
        "\t\t\"\"\")\n",
        "\texcept Exception as e:\n",
        "\t\treturn str(e)\n",
        "  \n",
        "@flask_app.route('/return_files/')\n",
        "def return_files():\n",
        "\ttry:\n",
        "\t\treturn send_file('./foo.txt', as_attachment=True)#, attachment_filename='python.jpg')\n",
        "\texcept Exception as e:\n",
        "\t\treturn str(e)\n",
        "\n",
        "\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    flask_app.run()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "INFO:__main__:Starting wormcells-de...\n",
            "INFO:werkzeug: * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            " * Running on http://dc39deefd37c.ngrok.io\n",
            " * Traffic stats available on http://127.0.0.1:4040\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "INFO:werkzeug:127.0.0.1 - - [22/Jan/2021 04:06:34] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [22/Jan/2021 04:06:35] \"\u001b[37mGET /static/js/clientside_table.js HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [22/Jan/2021 04:06:35] \"\u001b[37mGET /tables/ HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [22/Jan/2021 04:06:36] \"\u001b[37mGET /tables/ HTTP/1.1\u001b[0m\" 200 -\n",
            "INFO:werkzeug:127.0.0.1 - - [22/Jan/2021 04:06:36] \"\u001b[37mGET /static/favicon.ico HTTP/1.1\u001b[0m\" 200 -\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sEvpUCviqQ_o"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}